README.md
Rack::DeviseCookieAuth [![Build Status](https://secure.travis-ci.org/jilion/rack-devise_cookie_auth.png?branch=master)](http://travis-ci.org/jilion/rack-devise_cookie_auth) [![Dependency Status](https://gemnasium.com/jilion/rack-devise_cookie_auth.png)](https://gemnasium.com/jilion/rack-devise_cookie_auth) [![Code Climate](https://codeclimate.com/badge.png)](https://codeclimate.com/github/jilion/rack-devise_cookie_auth)
=====

Rack::DeviseCookieAuth allows to log-in from a [Devise](https://github.com/plataformatec/devise) remember-me token stored in a cookie.

It depends on Active::Support >= 4.0.0 and is tested against Ruby 1.9.3, 2.0.0 and the latest versions of Rubinius & JRuby.

## Installation

Add this line to your application's Gemfile (probably in the `:production` group only):

```ruby
gem 'rack-devise_cookie_auth'
```

And then execute:

```
$ bundle
```

## Basic Usage

If you don't use Bundler, be sure to require Rack::DeviseCookieAuth manually before actually using the middleware:

```ruby
 require 'rack/devise_cookie_auth'
 use Rack::DeviseCookieAuth, secret: 'YOUR_SESSION_SECRET'
```

To use Rack::DeviseCookieAuth in your Rails application, add the following line to your application config file (`config/application.rb` or `config/environments/production.rb`) for Rails 4:

```ruby
config.middleware.use Rack::DeviseCookieAuth, secret: 'YOUR_SESSION_SECRET'

# or if you're using Rack::Cache, be sure to insert Rack::DeviseCookieAuth before
config.middleware.insert_before Rack::Cache, Rack::DeviseCookieAuth, secret: 'YOUR_SESSION_SECRET'
```

Please note that the `:secret` option is mandatory and can normally be found in the 'config/initializers/secret_token.rb' file of your Rails app.

## Options

### Resource name

By default, the middleware will look for the user resource cookie named "remember_user_token" but if want to use authenticate againts another resource name you can set it with the `:resource` option:

```ruby
config.middleware.use Rack::DeviseCookieAuth, secret: 'YOUR_SESSION_SECRET', resource: 'admin'
```

### Redirection URL

By default, the middleware will redirect to the root path of the current domain but you can customize the redirection path with the `:redirect_to` option:

```ruby
# It can be a path...
config.middleware.use Rack::DeviseCookieAuth, secret: 'YOUR_SESSION_SECRET', redirect_to: '/login'

# ... or a full URL
config.middleware.use Rack::DeviseCookieAuth, secret: 'YOUR_SESSION_SECRET', redirect_to: 'https://yourdomain.com/login'
```

Development
-----------

* Documentation hosted at [RubyDoc](http://rubydoc.info/github/jilion/rack-devise_cookie_auth/master/frames).
* Source hosted at [GitHub](https://github.com/jilion/rack-devise_cookie_auth).

Pull requests are very welcome! Please try to follow these simple rules if applicable:

* Please create a topic branch for every separate change you make.
* Make sure your patches are well tested.
* Update the [README](https://github.com/jilion/rack-devise_cookie_auth/blob/master/README.md).
* Update the [CHANGELOG](https://github.com/jilion/rack-devise_cookie_auth/blob/master/CHANGELOG.md) for noteworthy changes.
* Please **do not change** the version number.

### Authors

* [Rémy Coutable](https://github.com/rymai) ([@rymai](http://twitter.com/rymai), [rymai.me](http://rymai.me))
* [Thibaud Guillaume-Gentil](https://github.com/thibaudgg) ([@thibaudgg](http://twitter.com/thibaudgg), [thibaud.me](http://thibaud.me))

### Contributors

[https://github.com/jilion/rack-devise_cookie_auth/graphs/contributors](https://github.com/jilion/rack-devise_cookie_auth/contributors)


CHANGELOG.md
## 1.0.0 - August 29, 2013

- Support Rails 4 (only)

## 0.1.0 - January 9, 2013

### Initial release

<!--- The following link definition list is generated by PimpMyChangelog --->
[@rymai]: https://github.com/rymai


test/test_helper.rb
require 'rubygems'
require 'test/unit'
require 'shoulda'
require 'rack/test'

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
require 'rack/devise_cookie_auth'

class Test::Unit::TestCase
  include Rack::Test::Methods

  def app
    Rack::Lint.new(@app)
  end

  def mock_app(options = {})
    @options = options
    @app     = app_builder
  end

  def verifier(secret = nil)
    key_generator = ActiveSupport::KeyGenerator.new(secret || @options[:secret], iterations: 1000)
    secret = key_generator.generate_key('signed cookie')
    ActiveSupport::MessageVerifier.new(secret)
  end

  private

  def app_builder
    builder = Rack::Builder.new
    builder.use Rack::DeviseCookieAuth, @options
    builder.run(main_app)
    builder.to_app
  end

  def main_app
    lambda { |env|
      headers = { 'Content-Type' => 'text/html' }
      [200, headers, ['Hello world!']]
    }
  end

  def set_cookie!(name, value)
    set_cookie("#{name}=#{value}; expires=#{Time.now + 3600}; secure; HttpOnly", URI.parse("https://www.example.org"))
  end

end


test/rack/cookie_auth_test.rb
require 'test_helper'

class TestRackCookieAuth < Test::Unit::TestCase

  context 'no cookie secret passed' do
    should 'raise an argument error' do
      assert_raise ArgumentError do
        mock_app
      end
    end
  end

  context 'no cookie' do
    setup {
      mock_app cookie_secret: 'abcd1234'
      expires_at = Time.now + 3600
    }

    should 'redirect to root' do
      get 'https://www.example.org/admin'
      assert_equal 302, last_response.status
      assert_equal 'https://www.example.org?user_return_to=https://www.example.org/admin', last_response.location
    end
  end

  context 'valid cookie secret passed' do
    setup {
      @expires_at = Time.now + 3600
    }

    context 'no options' do
      setup {
        mock_app cookie_secret: 'abcd1234'
        set_cookie!('remember_user_token', verifier.generate([1, @expires_at]), @expires_at)
      }

      should 'succeeds' do
        get 'https://www.example.org/admin'
        assert_equal 200, last_response.status
      end
    end

    context ':cookie_name option' do
      setup {
        mock_app cookie_secret: 'abcd1234', cookie_name: 'admin_remember_token'
        set_cookie!('admin_remember_token', verifier.generate([1, @expires_at]), @expires_at)
      }

      should 'succeeds' do
        get 'https://www.example.org/admin'
        assert_equal 200, last_response.status
      end
    end

    context 'invalid cookie' do
      context 'no options' do
        setup {
          mock_app cookie_secret: 'abcd1234'
          set_cookie!('remember_user_token', verifier('1234abcd').generate([1, @expires_at]), @expires_at)
        }

        should 'redirect to root' do
          get 'https://www.example.org/admin'
          assert_equal 302, last_response.status
          assert_equal 'https://www.example.org?user_return_to=https://www.example.org/admin', last_response.location
        end
      end

      context ':cookie_name option' do
        setup {
          mock_app cookie_secret: 'abcd1234', cookie_name: 'admin_remember_token'
          set_cookie!('admin_remember_token', verifier('1234abcd').generate([1, @expires_at]), @expires_at)
        }

        should 'redirect to given :redirect_to' do
          get 'https://www.example.org/admin'
          assert_equal 302, last_response.status
          assert_equal 'https://www.example.org?user_return_to=https://www.example.org/admin', last_response.location
        end
      end

      context ':redirect_to option' do
        context 'full URL' do
          setup {
            mock_app cookie_secret: 'abcd1234', redirect_to: 'https://www.example.org/login'
            set_cookie!('admin_remember_token', verifier('1234abcd').generate([1, @expires_at]), @expires_at)
          }

          should 'redirect to given :redirect_to' do
            get 'https://www.example.org/admin'
            assert_equal 302, last_response.status
            assert_equal 'https://www.example.org/login?user_return_to=https://www.example.org/admin', last_response.location
          end
        end

        context 'path without initial /' do
          setup {
            mock_app cookie_secret: 'abcd1234', redirect_to: 'login'
            set_cookie!('admin_remember_token', verifier('1234abcd').generate([1, @expires_at]), @expires_at)
          }

          should 'redirect to given :redirect_to' do
            get 'https://www.example.org/admin'
            assert_equal 302, last_response.status
            assert_equal 'https://www.example.org/login?user_return_to=https://www.example.org/admin', last_response.location
          end
        end

        context 'path with initial /' do
          setup {
            mock_app cookie_secret: 'abcd1234', redirect_to: '/login'
            set_cookie!('admin_remember_token', verifier('1234abcd').generate([1, @expires_at]), @expires_at)
          }

          should 'redirect to given :redirect_to' do
            get 'https://www.example.org/admin'
            assert_equal 302, last_response.status
            assert_equal 'https://www.example.org/login?user_return_to=https://www.example.org/admin', last_response.location
          end
        end
      end

      context ':return_to_param_name option' do
        context 'is custom' do
          setup {
            mock_app cookie_secret: 'abcd1234', redirect_to: 'https://www.example.org/login', return_to_param_key: 'admin_return_to'
            set_cookie!('admin_remember_token', verifier('1234abcd').generate([1, @expires_at]), @expires_at)
          }

          should 'not add return_to params' do
            get 'https://www.example.org/admin'
            assert_equal 302, last_response.status
            assert_equal 'https://www.example.org/login?admin_return_to=https://www.example.org/admin', last_response.location
          end
        end

        context 'is nil' do
          setup {
            mock_app cookie_secret: 'abcd1234', redirect_to: 'https://www.example.org/login', return_to_param_key: nil
            set_cookie!('admin_remember_token', verifier('1234abcd').generate([1, @expires_at]), @expires_at)
          }

          should 'not add return_to params' do
            get 'https://www.example.org/admin'
            assert_equal 302, last_response.status
            assert_equal 'https://www.example.org/login', last_response.location
          end
        end
      end
    end

    context 'expired cookie' do
      setup {
        @expires_at = Time.now - 3600
      }

      context 'no options' do
        setup {
          mock_app cookie_secret: 'abcd1234'
          set_cookie!('remember_user_token', verifier.generate([1, @expires_at]), @expires_at)
        }

        should 'redirect to root' do
          get 'https://www.example.org/admin'
          assert_equal 302, last_response.status
          assert_equal 'https://www.example.org?user_return_to=https://www.example.org/admin', last_response.location
        end
      end

      context ':cookie_name option' do
        setup {
          mock_app cookie_secret: 'abcd1234', cookie_name: 'admin_remember_token'
          set_cookie!('admin_remember_token', verifier.generate([1, @expires_at]), @expires_at)
        }

        should 'redirect to given :redirect_to' do
          get 'https://www.example.org/admin'
          assert_equal 302, last_response.status
          assert_equal 'https://www.example.org?user_return_to=https://www.example.org/admin', last_response.location
        end
      end

      context ':redirect_to option' do
        setup {
          mock_app cookie_secret: 'abcd1234', redirect_to: 'https://www.example.org/login'
          set_cookie!('admin_remember_token', verifier.generate([1, @expires_at]), @expires_at)
        }

        should 'redirect to given :redirect_to' do
          get 'https://www.example.org/admin'
          assert_equal 302, last_response.status
          assert_equal 'https://www.example.org/login?user_return_to=https://www.example.org/admin', last_response.location
        end
      end
    end

  end
end


test/rack/devise_cookie_auth_test.rb
require 'test_helper'

class TestRackDeviseCookieAuth < Test::Unit::TestCase

  context 'no cookie secret passed' do
    should 'raise an argument error' do
      assert_raise ArgumentError do
        mock_app
      end
    end
  end

  context 'no cookie' do
    setup {
      mock_app secret: 'abcd1234'
    }

    should 'redirect to root' do
      get 'https://www.example.org/admin'
      assert_equal 302, last_response.status
      assert_equal 'https://www.example.org?user_return_to=https://www.example.org/admin', last_response.location
    end
  end

  context 'valid cookie secret passed' do
    context 'no options' do
      setup {
        mock_app secret: 'abcd1234'
        set_cookie!('remember_user_token', verifier.generate([[1], 'remember_key']))
      }

      should 'succeeds' do
        get 'https://www.example.org/admin'
        assert_equal 200, last_response.status
      end

      should 'sets current_user_id env' do
        get 'https://www.example.org/admin'
        assert_equal 1, last_request.env['current_user_id']
      end
    end

    context ':resource option' do
      setup {
        mock_app secret: 'abcd1234', resource: 'admin'
        set_cookie!('remember_admin_token', verifier.generate([[1], 'remember_key']))
      }

      should 'succeeds' do
        get 'https://www.example.org/admin'
        assert_equal 200, last_response.status
      end

      should 'sets current_user_id env' do
        get 'https://www.example.org/admin'
        assert_equal 1, last_request.env['current_admin_id']
      end
    end
  end

  context 'invalid cookie' do
    context 'no options' do
      setup {
        mock_app secret: 'abcd1234'
        set_cookie!('remember_user_token', verifier('1234abcd').generate([[1], 'remember_key']))
      }

      should 'redirect to root' do
        get 'https://www.example.org/admin'
        assert_equal 302, last_response.status
        assert_equal 'https://www.example.org?user_return_to=https://www.example.org/admin', last_response.location
      end
    end

    context ':resource option' do
      setup {
      mock_app secret: 'abcd1234', resource: 'admin'
        set_cookie!('remember_admin_token', verifier('1234abcd').generate([[1], 'remember_key']))
      }

      should 'redirect to root' do
        get 'https://www.example.org/admin'
        assert_equal 302, last_response.status
        assert_equal 'https://www.example.org?admin_return_to=https://www.example.org/admin', last_response.location
      end
    end

    context ':redirect_to option' do
      context 'full URL' do
        setup {
          mock_app secret: 'abcd1234', redirect_to: 'https://www.example.net/login'
          set_cookie!('remember_user_token', verifier('1234abcd').generate([[1], 'remember_key']))
        }

        should 'redirect to given :redirect_to' do
          get 'https://www.example.org/admin'
          assert_equal 302, last_response.status
          assert_equal 'https://www.example.net/login?user_return_to=https://www.example.org/admin', last_response.location
        end
      end

      context 'path without initial /' do
        setup {
          mock_app secret: 'abcd1234', redirect_to: 'login'
          set_cookie!('remember_user_token', verifier('1234abcd').generate([[1], 'remember_key']))
        }

        should 'redirect to given :redirect_to' do
          get 'https://www.example.org/admin'
          assert_equal 302, last_response.status
          assert_equal 'https://www.example.org/login?user_return_to=https://www.example.org/admin', last_response.location
        end
      end

      context 'path with initial /' do
        setup {
          mock_app secret: 'abcd1234', redirect_to: '/login'
          set_cookie!('remember_user_token', verifier('1234abcd').generate([[1], 'remember_key']))
        }

        should 'redirect to given :redirect_to' do
          get 'https://www.example.org/admin'
          assert_equal 302, last_response.status
          assert_equal 'https://www.example.org/login?user_return_to=https://www.example.org/admin', last_response.location
        end
      end
    end
  end
end


lib/rack/cookie_auth.rb
require 'rack/cookie_auth/version'
require 'active_support/message_verifier'

module Rack
  class CookieAuth
    DEFAULT_OPTIONS = {
      cookie_name: 'remember_user_token',
      user_id_key: 'current_user_id',
      redirect_to: nil,
      return_to_param_key: 'user_return_to'
    }
    def initialize(app, options = {})
      raise ArgumentError, 'Cookie secret must be set!' if options[:cookie_secret].nil?

      @app, @options = app, DEFAULT_OPTIONS.merge(options)
    end

    def call(env)
      @request = Rack::Request.new(env)

      verifier = ActiveSupport::MessageVerifier.new(@options[:cookie_secret])
      env[@options[:user_id_key]], expires_at = verifier.verify(@request.cookies[@options[:cookie_name]])

      if expires_at > Time.now
        @app.call(env)
      else
        redirect!
      end
    rescue ActiveSupport::MessageVerifier::InvalidSignature
      redirect!
    end

    private

    def redirect!
      [302, { 'Content-Type' => 'text/html', 'Location' => redirect_url }, ["Redirected to #{redirect_url}!"]]
    end

    def redirect_url
      [redirect_to, return_to].compact.join('?')
    end

    def redirect_to
      if @options[:redirect_to] =~ %r{\Ahttps?://}
        @options[:redirect_to]
      else
        @request.url.sub(@request.fullpath, "/#{@options[:redirect_to]}".squeeze('/').sub(%r{/\z}, '') || '')
      end
    end

    def return_to
      if @options[:return_to_param_key]
        "#{@options[:return_to_param_key]}=#{@request.url}"
      end
    end

  end
end


lib/rack/devise_cookie_auth.rb
require 'rack/devise_cookie_auth/version'
require 'active_support/key_generator'
require 'active_support/message_verifier'

module Rack
  class DeviseCookieAuth
    DEFAULT_OPTIONS = {
      resource: 'user',
      redirect_to: nil
    }
    def initialize(app, options = {})
      raise ArgumentError, 'Cookie secret must be set!' if options[:secret].nil?

      @app, @options = app, DEFAULT_OPTIONS.merge(options)
    end

    def call(env)
      @request = Rack::Request.new(env)

      resource_ids, remember_key = verifier.verify(@request.cookies[cookie_name])
      env["current_#{resource}_id"] = resource_ids.first

      @app.call(env)
    rescue ActiveSupport::MessageVerifier::InvalidSignature
      redirect!
    end

    private

    def redirect!
      [302, { 'Content-Type' => 'text/html', 'Location' => redirect_url }, ["Redirected to #{redirect_url}!"]]
    end

    def verifier
      @verifier ||= ActiveSupport::MessageVerifier.new(secret)
    end

    def secret
      key_generator = ActiveSupport::KeyGenerator.new(@options[:secret], iterations: 1000)
      key_generator.generate_key('signed cookie')
    end

    def resource
      @options[:resource].to_s
    end

    def cookie_name
      "remember_#{resource}_token"
    end

    def redirect_url
      [redirect_to, return_to].compact.join('?')
    end

    def redirect_to
      if @options[:redirect_to] =~ %r{\Ahttps?://}
        @options[:redirect_to]
      else
        @request.url.sub(@request.fullpath, "/#{@options[:redirect_to]}".squeeze('/').sub(%r{/\z}, '') || '')
      end
    end

    def return_to
      "#{resource}_return_to=#{@request.url}"
    end

  end
end
